File: bot/lib/db/teams.py
from typing import TypedDict
from bot.lib.convex_client import client
from bot.lib.log import log
import time


class TeamDto(TypedDict):
    _id: str
    _creationTime: float
    name: str
    captainId: str
    hasFirstPick: bool
    threadId: str | None
    voiceChannelId: str | None
    updateTime: int


class TeamsServiceImpl:

    @staticmethod
    def find_by_id(team_id: str) -> TeamDto | None:
        try:
            result: TeamDto | None = client.query("teams:findById", {"teamId": team_id})
            return result
        except Exception as e:
            log(f"Error fetching team {team_id}: {e}")
            raise

    @staticmethod
    def find_by_thread_id(thread_id: str) -> TeamDto | None:
        try:
            result: TeamDto | None = client.query("teams:findByThreadId", {"threadId": thread_id})
            return result
        except Exception as e:
            log(f"Error fetching team by thread {thread_id}: {e}")
            raise

    @staticmethod
    def create(name: str, captain_id: str, has_first_pick: bool) -> str:
        try:
            result: str = client.mutation(
                "teams:create",
                {
                    "name": name,
                    "captainId": captain_id,
                    "hasFirstPick": has_first_pick,
                    "updateTime": int(time.time() * 1000)
                }
            )
            log(f"Team '{name}' created with captain {captain_id}")
            return result
        except Exception as e:
            log(f"Error creating team {name}: {e}")
            raise

File: convex/teams.ts
import { mutation, query } from "./_generated/server";
import { v } from "convex/values";

export const findById = query({
  args: { teamId: v.id("teams") },
  async handler(ctx, args) {
    return await ctx.db.get(args.teamId);
  },
});

export const findByThreadId = query({
  args: { threadId: v.string() },
  async handler(ctx, args) {
    return await ctx.db
      .query("teams")
      .filter((q) => q.eq(q.field("threadId"), args.threadId))
      .first();
  },
});

export const create = mutation({
  args: {
    name: v.string(),
    captainId: v.id("users"),
    hasFirstPick: v.boolean(),
    updateTime: v.number(),
  },
  async handler(ctx, args) {
    const teamId = await ctx.db.insert("teams", {
      name: args.name,
      captainId: args.captainId,
      hasFirstPick: args.hasFirstPick,
      updateTime: args.updateTime,
    });

    return teamId;
  },
});

export const updateThreadId = mutation({
  args: {
    teamId: v.id("teams"),
    threadId: v.string(),
    updateTime: v.number(),
  },
  async handler(ctx, args) {
    await ctx.db.patch(args.teamId, {
      threadId: args.threadId,
      updateTime: args.updateTime,
    });
  },
});

export const updateVoiceChannelId = mutation({
  args: {
    teamId: v.id("teams"),
    voiceChannelId: v.string(),
    updateTime: v.number(),
  },
  async handler(ctx, args) {
    await ctx.db.patch(args.teamId, {
      voiceChannelId: args.voiceChannelId,
      updateTime: args.updateTime,
    });
  },
});
